### PR TITLE
Remove unloaded resources from task queues

### LINKED ISSUE #2732: Layout promises never resolve when elements are destroyed before load fired
If the user load an article with say Instagram embeds and then swipe to move the the next one quickly and then swipe back to that article the Instagram embed will never load if the instagram layout promise never resolved the first time.

The reason this happen is that runtime will never schedule the element for a layout again because the first layout is still in our `exec` queue and will never exit because the first `loadPromise` will never resolve or error.

This is specially critical on slower connections where resolving the load promise for iframes might take along time and users navigate away from the document before that happens.

## How do we reproduce the issue?
I haven't tried to reproduce this on search carousel, but in theory it should be reproducible there. I was able to reproduce this issue on the examples/viewer.html setup with `amp-instagram.html` example set as the 2nd viewer.
1. In the viewer.html navigate to 2nd (instagram example) viewer and quickly navigate back to another one
2. This will force unlayout for all instagram elements 
3. Now go back to 2nd viewer and notice instagram posts will never load
4. If you look in the console, the `exec` queue will always have the same number of never "ending" tasks

I'll try to reproduce this on carousel as well. But in general we need to account for removing DOM before the layout promise resolve.

### PR DESCRIPTION
**Pull request by mkhatib**
Fixes #2732 

cc/ @jridgewell

### CODE REVIEW

File: extensions/amp-brightcove/0.1/amp-brightcove.js [hunk 1]
```diff
@@ -121,7 +121,7 @@ class AmpBrightcove extends AMP.BaseElement {
    *
    * See https://github.com/ampproject/amphtml/issues/2224 for information.
    */
-  unlayout() {
+  unlayoutCallback() {
```
Comment (jridgewell) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Whoops.

**Comment by jridgewell** [comment]
LGTM, but I defer to @dvoytenko.

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: @mkhatib This should have also `@override` jsdoc annotation. Please separate this into a separate PR and merge ASAP - no need to go through review.

File: src/service/resources-impl.js [hunk 2]
```diff
@@ -1402,6 +1408,37 @@ export class Resources {
       // Selection API not supported.
     }
   }
+
+  /**
+   * Cleanup task queues from tasks for elements that has been unloaded.
+   * @param resource
+   * @private
+   */
+  maybeRemoveFromQueues_(resource) {
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Don't see "maybe" part here :) It looks very definitive to me.

File: src/service/resources-impl.js [hunk 3]
```diff
@@ -1402,6 +1408,37 @@ export class Resources {
...
+    if (resource.getState() == ResourceState_.NOT_LAID_OUT) {
+      // If the layout promise for this resource has not resolved yet, remove
+      // it from the task queues to make sure this resource can be rescheduled
+      // for layout again later on.
+      // TODO(mkhatib): Look into if we need to do the same for preload tasks, this
+      // depends on whether we expect people to unload prerendered DOM.
+      // The answer might be no since we probably want to keep the prerendered
+      // DOM around to use as placeholder when the user navigate back to the
+      // document.
+      const taskId = resource.getTaskId(LAYOUT_TASK_ID_);
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: As number of types of tasks grows - this will be less and less tenable. I think we should just iterate through queues and remove found entries via `resource == task.resource`. The queues should be small - that's the point. If we have concerns about the performance then the first best place would be to always add API to TaskQueue: `.dequeueByElement()` or such where we can freely optimize the implementation and unit-test separately.

File: src/service/resources-impl.js [hunk 4]
```diff
@@ -1402,6 +1408,37 @@ export class Resources {
...
+            'layout promise ejected from queue after unload',
+            resource.debugid);
+      }
+      if (executingTask && this.exec_.dequeue(executingTask)) {
+        dev.fine(TAG_,
+            'layout promise ejected from exec queue after unload',
+            resource.debugid);
+      }
+    }
+  }
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: Please cleanup the `requestsChangeSize_` queue as well.

Comment (mkhatib) [inline src/service/resources-impl.js]: I was looking for that queue. But it seems we're not using the `TaskQueue_` implementation for that one. It seems that it's just a normal JS array, is that correct?

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Sent #2766 and added a test case

File: src/service/resources-impl.js (hunk 2 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: So the maybe part is two parts:
- Has the resource been unlaid out (by the indication of NOT_LAID_OUT) - This is only changed when the unlayoutCallback actually returns true. 
- Is the task actually in the queue?

I wasn't sure what to call it. Maybe rename it to something like `removeStaledTasks`, `cleanupQueuesFromStaledTasks`?

File: src/service/resources-impl.js (hunk 4 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done

File: src/service/resources-impl.js (hunk 3 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done. I've implemented a `.filter` method - similar to the native `.filter` for arrays with some differences (looping in reverse order because we're manipulating the array and calling this.dequeue on matches - which would remove the tasks).

**Comment by mkhatib** [comment]
PTAL

File: src/service/resources-impl.js [hunk 5]
```diff
@@ -1402,6 +1408,29 @@ export class Resources {
...
+      // for layout again later on.
+      // TODO(mkhatib): Think about how this might affect preload tasks once the
+      // prerender change is in.
+      this.queue_.filter(task => {
+        return task.resource != resource;
+      });
+      this.exec_.filter(task => {
+        return task.resource != resource;
+      });
+      this.requestsChangeSize_ =  this.requestsChangeSize_.filter(
```
Comment (jridgewell) [inline src/service/resources-impl.js]: Why are we assigning in this one, but not the others?

Comment (mkhatib) [inline src/service/resources-impl.js]: This is not really a TaskQueue_ instance it's just a JavaScript array. Native .filter doesn't run in place on the array but returns a new array with the filtered elements.

File: src/service/resources-impl.js (hunk 2 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: `cleanupTasks`?

File: src/service/resources-impl.js (hunk 4 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: Correct. Why? Not sure.

File: src/service/resources-impl.js (hunk 5 above)
Comment (dvoytenko) [inline src/service/resources-impl.js]: Let's rename the tasks's API then. Maybe let's call it `purge`? Or `cleanup` or such. But using `filter` is definitely misleading.

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: So, this is gone, right?

File: test/functional/test-resources.js [hunk 6]
```diff
@@ -26,7 +26,7 @@ import * as sinon from 'sinon';
 
 /*eslint "google-camelcase/google-camelcase": 0*/
 describe('Resources', () => {
-
+  //
```
Comment (dvoytenko) [inline test/functional/test-resources.js]: ?

File: src/service/resources-impl.js (hunk 5 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: test/functional/test-resources.js (hunk 6 above)
Comment (mkhatib) [inline test/functional/test-resources.js]: Gone.

File: src/service/resources-impl.js (hunk 2 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

File: extensions/amp-brightcove/0.1/amp-brightcove.js (hunk 1 above)
Comment (mkhatib) [inline extensions/amp-brightcove/0.1/amp-brightcove.js]: Gone.

**Comment by mkhatib** [comment]
PTAL

File: src/service/resources-impl.js [hunk 7]
```diff
@@ -2052,6 +2091,19 @@ export class TaskQueue_ {
...
+
+  /**
+   * Dequeue tasks in queue removing any task that pass the callback test.
+   * @param {function(!TaskDef)} callback Return true to dequeue the task.
+   */
+  purge(callback) {
+    let index = this.tasks_.length;
+    while (index--) {
+      if (callback(this.tasks_[index])) {
+        this.dequeue(this.tasks_[index]);
```
Comment (jridgewell) [inline src/service/resources-impl.js]: This is still an n<sup>2</sup> algorithm because of the `#indexOf` search inside `#dequeue`. Is that an issue @dvoytenko?

Comment (dvoytenko) [inline src/service/resources-impl.js]: Yeah, let's get rid of the second loop since we already have all indexes we need. A single loop is fine though given that removals are rare.

Comment (mkhatib) [inline src/service/resources-impl.js]: Done.

**Comment by mkhatib** [comment]
PTAL

File: src/service/resources-impl.js [hunk 8]
```diff
@@ -2053,6 +2091,36 @@ export class TaskQueue_ {
...
+      return false;
+    }
+    this.tasks_.splice(index, 1);
+    delete this.taskIdMap_[task.id];
+    return true;
+  }
+
+  /**
+   * Dequeue tasks in queue removing any task that pass the callback test.
+   * @param {function(!TaskDef)} callback Return true to dequeue the task.
```
Comment (dvoytenko) [inline src/service/resources-impl.js]: `function(...):boolean`

**Comment by dvoytenko** [comment]
LGTM with one more nit

File: src/service/resources-impl.js (hunk 8 above)
Comment (mkhatib) [inline src/service/resources-impl.js]: Done.
